### PR TITLE
 GenerateToNotebook  and Deploy buttons for Notebook Wizards

### LINKED ISSUE #12577: Move Arc instance deployments to use Notebook wizards
It's a better deployment experience for now and lets users configure settings prior to deploying.

Should also make Notebook wizards support "Script" and "Run" buttons - script will generate notebook but not execute while run would script and execute notebook.

[AB#3777](https://mssqltools.visualstudio.com/dea8be32-0532-4a8f-a2f4-89271d8b1cc5/_workitems/edit/3777)

### PR DESCRIPTION
**Pull request by ranasaria**
This change adds a feature to support both ScriptToNotebook and Deploy buttons in a Notebook Wizard. The Deploy button now runs the notebook instead of executing the notebook using azdata in a new process.

Testing:
1. End to end testing of the new feature.
2. Extension tests.
3. Ensuring no regression on BDC Wizard.

This PR fixes #12577

### CODE REVIEW

File: extensions/resource-deployment/src/ui/notebookWizard/notebookWizard.ts [hunk 1]
```diff
@@ -49,15 +48,32 @@ export class NotebookWizard extends WizardBase<NotebookWizard, NotebookWizardPag
...
-		this.wizardObject.doneButton.label = this.wizardInfo.actionText;
 	}
 
 	protected onCancel(): void {
 	}
 
+	protected async onGenerateScript(): Promise<void> {
+		const { notebook, env } = await this.prepareNotebookAndEnvironment();
+		try {
+			Object.assign(process.env, env);
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/notebookWizard/notebookWizard.ts:59]: Can you check whether this is going to be an issue for scripting multiple deployment notebooks? Specifically this scenario :

1. Script MIAA deployment A to Notebook, but don't execute it
2. Script MIAA deployment B to Notebook with different env vars (password is an easy one), but don't execute it
3. Go back and execute MIAA deployment Notebook A

Will A still be using its own values or will it have been overwritten with B's?

If it is overwritten we could solve that by appending an identifier onto the env var names - just do a counter that starts at 1 and gets incremented every time a Notebook is deployed (and do a search/replace in the Notebook prior to scripting)

If you do have to do that let's do it as a separate PR so as to not clutter up this one.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: extensions/resource-deployment/src/ui/notebookWizard/notebookWizard.ts (hunk 1 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/ui/notebookWizard/notebookWizard.ts:59]: Like your suggestion. I will do this test. However, GenerateScript is the default action in main currently so this issue/test can be addressed independently of this PR.

File: extensions/arc/package.json [hunk 2]
```diff
@@ -144,9 +144,8 @@
             "notebookWizard": {
               "notebook": "./notebooks/arcDeployment/deploy.arc.data.controller.ipynb",
               "type": "new-arc-control-plane",
-              "runNotebook": false,
+              "runNotebook": "userChooses",
```
Comment (alanrenmsft) [inline extensions/arc/package.json:147]: can you come up with a better name here? 😄

Comment (Charles-Gagnon) [reply extensions/arc/package.json:147]: FWIW I don't have a huge issue with the name as it currently is - but I do slightly prefer that we keep the type the same (so some other string values besides boolean true and false)

Comment (alanrenmsft) [reply extensions/arc/package.json:147]: it should an adjective here, right? how about 'userSelected'?

Comment (ranasaria) [reply extensions/arc/package.json:147]: I suck at naming. I would appreciate suggestions.

File: extensions/resource-deployment/src/services/notebookService.ts [hunk 3]
```diff
@@ -50,7 +50,7 @@ export class NotebookService implements INotebookService {
 	 * Launch notebook with file path
 	 * @param notebook the path of the notebook
 	 */
-	async launchNotebook(notebook: string | NotebookPathInfo): Promise<azdata.nb.NotebookEditor> {
+	async openNotebook(notebook: string | NotebookPathInfo): Promise<azdata.nb.NotebookEditor> {
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/services/notebookService.ts:53]: Update doc comment

File: extensions/resource-deployment/src/services/notebookService.ts [hunk 4]
```diff
@@ -78,7 +78,7 @@ export class NotebookService implements INotebookService {
 	 * @param title the title of the notebook
 	 * @param content the notebook content
 	 */
-	async launchNotebookWithContent(title: string, content: string): Promise<azdata.nb.NotebookEditor> {
+	async openNotebookWithContent(title: string, content: string): Promise<azdata.nb.NotebookEditor> {
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/services/notebookService.ts:81]: Update doc comments

File: extensions/arc/package.json (hunk 2 above)
Comment (Charles-Gagnon) [reply extensions/arc/package.json:147]: This is probably fine for now - eventually we may want to allow a bit more flexibility (have more types of actions) but I'm not too worried about making this super generic right now.

I would suggest removing the enabled property and then just having the non-OK options be optional so if they aren't included they won't show up.

Comment (ranasaria) [reply extensions/arc/package.json:147]: Sounds good. So, then the doneAction is always defined so a user will always have the option of running the notebook and optionally the wizard can add a scriptAction to allow for just scripting of the notebook.

File: extensions/resource-deployment/src/ui/notebookWizard/notebookWizard.ts (hunk 1 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/ui/notebookWizard/notebookWizard.ts]: This test is complete. The passwords work as expected, so we do not need to make any additional changes.

**Review by Charles-Gagnon — APPROVED** [review]

File: extensions/resource-deployment/src/interfaces.ts [hunk 5]
```diff
@@ -114,21 +114,57 @@ export interface BdcWizardInfo {
...
 	codeCellInsertionPosition?: number;
+	/**
+	 * 	This array defines the json for the pages of this wizard.
+	 */
 	pages: NotebookWizardPageInfo[]
 }
 
 export interface WizardInfoBase extends FieldInfoBase {
+	/**
+	 * 	the taskName to use when running the notebook as a background task. The default value is the @see title of the wizard.
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/resource-deployment/src/interfaces.ts:144]: Will check to ensure none of subclasses are using it.

File: extensions/resource-deployment/src/interfaces.ts [hunk 6]
```diff
@@ -114,21 +114,57 @@ export interface BdcWizardInfo {
...
+	 */
 	taskName?: string;
 	type?: DeploymentType;
-	actionText?: string;
+	/**
+	 * 	done button attributes.
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/resource-deployment/src/interfaces.ts:149]: I do not see the extra space.

**Pull request merged**